### ISSUE TITLE
Array.reduce Array.append doesn't work 

### ISSUE THREAD

**Issue by MangelMaxime** [fable-compiler/Fable#2372]
Hello :)

### Repro code

```fs
let arr1 =
    [|
        1 
    |]

let arr2 =
    [|
        2 
    |]

let array2d =
    [|
        [| arr1 |]
        [| arr2 |]
    |]

let res = Array.reduce Array.append array2d
```

Produces the following error when executing:

```
TypeError: 1 is not a constructor
```

### Expected and actual results

Should not throw an error and gives back:

```fsi
val res : int [] [] = [|[|1|]; [|2|]|]
```

### Related information

* Fable version: 3.1.2
* Operating system: Windows 10

**Comment by inosik** [comment]
`Array.reduce` is translated to [JavaScripts `Array.reduce` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce), which has (kinda) the following signature:

``` js
Array<T>.reduce(callbackfn: (previousValue: T, currentValue: T, currentIndex: number, array: T[]) => T): T
```

This is Fables `Array.reduce` function:

https://github.com/fable-compiler/Fable/blob/238ed9f180b156b67f2c5af94d673119b40a647e/src/fable-library/Array.fs#L128-L136

`Array.reduce` method gives us the index of the current item as the `cons` parameter, which explains the "TypeError: 1 is not a constructor" error.

This F# code:

``` fsharp
Array.reduce f xs
```

Should be translated to something like this:

``` diff
-xs.reduce(f)
+xs.reduce((acc, item) => f(acc, item))
```

BTW, `Seq.reduce` has the same bug, because [it special-cases arrays](https://github.com/fable-compiler/Fable/blob/238ed9f180b156b67f2c5af94d673119b40a647e/src/fable-library/Seq.ts#L370).

**Comment by inosik** [comment]
@MangelMaxime can you use `Array.concat` instead of `reduce` + `append`?

I think I remember that this pattern was pointed out by FSharpLint, I wonder why it doesn't now 🤔

**Comment by MangelMaxime** [comment]
> @MangelMaxime can you use Array.concat instead of reduce + append?

@inosik I indeed used `Array.concat` in the end ;) I think I kind of disabled `FSharpLint` from Ionide and also I am switching back and forth between `Ionide` and `F# language server` (which doesn't have `FSharpLint`) because of performance reason. I really like Ionide but it stills has performance issues when project becomes bigger... at least for me.

**Comment by inosik** [comment]
> This F# code:
> ``` fsharp
> Array.reduce f xs
> ```
> Should be translated to something like this:
> ``` diff
> -xs.reduce(f)
> +xs.reduce((acc, item) => f(acc, item))
> ```

We already do that, but only if `Array.append` requires a special `cons`, like a numeric array constructor. The following sample works correctly for `allNums`, but not for the others:

``` fsharp
let nums =
    [|
        [| 0 |]
        [| 1 |]
    |]
let allNums = Array.reduce Array.append nums

let nums2d =
    [|
        [| [| 0 |] |]
        [| [| 1 |] |]
    |]
let allNums2d = Array.reduce Array.append nums2d

let strs =
    [|
        [| "a" |]
        [| "b" |]
    |]
let allStrs = Array.reduce Array.append strs
```

[REPL](https://fable.io/repl/#?code=DYUwLgBMD2DmEAoCGAneAuC0BGArAlBALwCwAUBJRAMoCeAzmCALYB0AwtAHb3SisB1FAEsmAGWFcQEVLHLlQkLgFdm9YuSoQA2gB9NWqnogAGCLoC6Bwzt0QAjOasUqlheBnBgAOVXqiEACCKChItKwoIAAmygDG0sGh4UgADikgXFEQKmoKcJ4+fvJkitl+AExZpC6UetaGxsZmlk71Wo12ji1uNa0lHkhevmqVxEEhYRHRcQkTyWkZWTn0lXnwg4UjUcWljCj+9XW97XYAREinfTa1Z9iXPVo9pRvUYPtjiZORMfHjSaypdKZCB7ehrAqvfbkIA&html=DwCwLgtgNgfAUKApgQwCbwAQeBRZkbhgAOAtIgI4CuAlgG4C8A5AMID2AdmIl6QCoBPYoiYYAxp25dm3AB5gA9OGgBucSGQAnAM54GVMADNSADiaZs2sZprEwGbZrEMARAH03ACQDyAZT4eClA0AEbaCpooYmAAdMSabKhU0TScMRA0HDEAVtouMMAKVjZ2FsDFtvaOzu5efgFuQaHhkcjRpKhsEHEJSSlpGVm5+YUVpQhKKOgIIYkC4lDI2tquyMRkElzImYia+XBYWMCo9Bg0qK6IUBnaIKRrxC4LSysuVzd3DyMKJ3TwhbNUAJ-kpILAgA&css=Q)

**Comment by alfonsogarciacaro** [comment]
Thanks a lot @inosik! This was an unfortunate combination of the function unwrapping optimization and the array constructor injection. We already had a check for null arguments so I've changed it to check also for function type, I hope that's enough :)
